### DISCUSSION TITLE
Unable to get OpenTK.OpenAL to Detect openal32.dll within directory or embedded from openal-soft

### DISCUSSION THREAD

**Discussion by Mrcubix** [opentk/opentk#1642] [Q&A] [answered]
I have been trying to get this wrapper to work today and was unable to test what i even written since i have been getting this dll import error.

I or any users shouldn't have to install OpenAL to make my app work so i tried imbedding and/or setting the openAL dll as content, both options does not seem to affect the result.

 Any ideas?

**Comment by NogginBops** [comment]
Can you give some more details of where you've placed your dll and what the project setup is?

**Reply by Mrcubix** [reply]
I trying to use the openal-soft's dll, My first attempt at getting it working was with OpenTK 4.7.7 .
When i couldn't get it wotking, i tried the same with OpenTK 3, there, i'm getting another exception about a bad image.
This could be a clue to what the issue is, anyway here is the relevant part of my current project file:

```csproj
<ItemGroup>
  <EmbeddedResource Include="native/openal32.dll">
    <Link>openal32.dll</Link>
    <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    <Pack>true</Pack>
  </EmbeddedResource>
</ItemGroup>

<ItemGroup>
  <EmbeddedResource Include="ColorTiles.config">
    <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
  </EmbeddedResource>
</ItemGroup>
```

I have tried the same code above with `Content` & `None` too.
  
ColorTiles.config : 
```csproj
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <dllmap dll="AL" os="windows" target="openal32.dll"/>
  <dllmap dll="AL" os="linux" target="libopenal.so.1"/>
</configuration>
```

I did also try to link the the 2 architectures provided by openal-soft to the specific arch folder in question (x86 & x64), didn't  make much of a difference.
I won't lie that i lack experience with including native librairies and getting dllImport to work with them, made sure to look at various threads and repos trying to do the same but not including any particular lib files themselves too for OpenAL.

I however did see that some SDL related repos that were including SDL2 dlls and whose nuget package did work.

Other than the above, i tried placing the dll in the same folder as the executable, made sure it was also in the running directory, none worked.

**Comment by jvbsl** [comment, marked as answer]
The configuration only works with mono, and therefore would only work with OpenTK 3.

For the newer OpenTK you could debug into https://github.com/opentk/opentk/blob/2293609c1b65220e2200c29b4ee86cbcb2cdad3d/src/OpenAL/OpenTK.Audio.OpenAL/Native/ALLoader.cs#L42
Where part of the resolving process should happen.

Apart from that should the nativel openal32.dll probably at the output folder in `runtimes/win-x64/native/openal32.dll` for the 64bit openal version and `runtimes/win-x86/native/openal32.dll` where these the locations you tried?.

Apart from that in general a BadImageFormat-Exception happens when you try to load a library of the wrong architecture(e.g. trying to load a ARM library on x86, or more often trying to load a x86 assembly in a x86_64 context or vice versa).

**Reply by Mrcubix** [reply]
> Apart from that should the nativel openal32.dll probably at the output folder in runtimes/win-x64/native/openal32.dll for the 64bit openal version and runtimes/win-x86/native/openal32.dll where these the locations you tried?.

I did try the former as well as the published form for both arch.

> Apart from that should the nativel openal32.dll probably at the output folder in runtimes/win-x64/native/openal32.dll for the 64bit openal version and runtimes/win-x86/native/openal32.dll where these the locations you tried?.

I did see that in the various thread i've read, but that isn't the case as the "win64" dll threw the exact same exception in OpenTK 3.

> The configuration only works with mono, and therefore would only work with OpenTK 3.
> 
> For the newer OpenTK you could debug into
> 
> https://github.com/opentk/opentk/blob/2293609c1b65220e2200c29b4ee86cbcb2cdad3d/src/OpenAL/OpenTK.Audio.OpenAL/Native/ALLoader.cs#L42
> 
> 
> Where part of the resolving process should happen.

I'll do that and let you know

**Reply by Mrcubix** [reply]
![image](https://github.com/opentk/opentk/assets/39861216/dea215a7-3574-4b34-ab63-85a83757f451)

at first glance, it look like loading it failed when in the same folder as the running binary

![image](https://github.com/opentk/opentk/assets/39861216/c5208748-f1c9-4677-a8cc-715bf9209ca4)

as for the runtimes dir where other native libs are, same result

**Reply by jvbsl** [reply]
What about `searchPath`?

**Reply by Mrcubix** [reply]
stays null

**Reply by Mrcubix** [reply]
Managed ro make it work with the runtimes path somehow after restarting enough time, something probably decide to change.

**Comment by jvbsl** [comment]
> I did see that in the various thread i've read, but that isn't the case as the "win64" dll threw the exact same exception in OpenTK 3.

Sorry, I meant for the new OpenTK, the runtimes folder does not work for OpenTK 3, as it was only introduced with newer .Net.

**Reply by Mrcubix** [reply]
I try things in OpenTK 4 first, only then i try in OpenTK 3
